docs/explanation/input-streams.md

```diff
@@ -0,0 +1,267 @@
+
+**Purpose**: Scanning barcodes from pre-recorded video files.
+
+**How it works**:
+1. Creates a new `<video>` element
+2. Sets the `src` attribute to the video URL
+3. Waits for video metadata to load
+4. Triggers `canrecord` when dimensions are known
+
+**Key characteristics**:
+- `ended()` returns the video element's ended state
+- Supports seeking via `setCurrentTime()`
+- Video plays frame-by-frame during scanning
+
+**Configuration example**:
+```javascript
+inputStream: {
+    type: 'VideoStream',
+    src: '/path/to/video.mp4'
+}
+```
+
+### ImageStream
+
+**Purpose**: Scanning barcodes from static images or image sequences.
+
+**How it works**:
+1. Parses the image URL configuration
+2. Uses `ImageLoader` to fetch the image(s)
+3. Reads EXIF data to handle image orientation
+4. Calculates dimensions based on size config
+5. Triggers `canrecord` when image(s) are loaded
+
+**Key characteristics**:
+- Can process a single image or a sequence
+- Handles EXIF orientation automatically
+- `ended()` returns true after all images are processed
+- Used internally by `decodeSingle()`
+
+**Configuration example (single image)**:
+```javascript
+inputStream: {
+    type: 'ImageStream',
+    src: '/path/to/barcode.jpg',
+    sequence: false
+}
+```
+
+**Configuration example (image sequence)**:
+```javascript
+inputStream: {
+    type: 'ImageStream',
+    src: '/path/to/images/img_%d.jpg',  // %d is replaced with frame number
+    sequence: true,
+    length: 10  // Number of images
+}
+```
+
+## Race Conditions and Async Behavior
+
+Because initialization involves async operations (camera access, file loading), race conditions can occur if:
+
+1. **`stop()` is called during `init()`**: The `canrecord` event may fire after `stop()` has begun cleanup. Quagga2 handles this with an `initAborted` flag.
+
+2. **React StrictMode double-invocation**: StrictMode mounts, unmounts, and remounts components, causing rapid `init() → stop() → init()` sequences.
+
+3. **Component unmounting before camera ready**: User navigates away before `getUserMedia()` resolves.
+
+**Best practices to avoid issues**:
+
+```javascript
+useLayoutEffect(() => {
+    let cancelled = false;
+    
+    Quagga.init(config, (err) => {
+        if (cancelled) return;  // Ignore if unmounted
+        if (err) {
+            console.error(err);
+            return;
+        }
+        Quagga.start();
+    });
+    
+    return () => {
+        cancelled = true;
+        Quagga.stop();
+    };
+}, []);
+```
+
+## Source Code
+
+The input stream system is implemented in:
+
+- `src/input/input_stream/input_stream_browser.ts` - Browser stream implementations
+- `src/input/input_stream/input_stream.ts` - Node.js stream implementation
+- `src/input/input_stream/input_stream.d.ts` - TypeScript interface
+- `src/quagga/setupInputStream.ts` - Stream factory selection
+- `src/input/camera_access.ts` - Camera permission handling
+- `src/input/frame_grabber.js` - Frame extraction from streams
```

Comment:
don't forget about there being platform specific frame grabber code for node and browser